include/range/v3/range_fwd.hpp

```diff
@@ -25,6 +25,13 @@
 #include <range/v3/utility/static_const.hpp>
 #include <range/v3/version.hpp>
 
+#if __has_include(<ranges>)
+#  include <ranges>
+#  if __cpp_lib_ranges >= 201911
+#    define RANGE_V3_USE_STD_ENABLE_VIEW 1
+#  endif
```

Comment:
I _really_ don't want to pull in all of `<ranges>`. I'm not sure how else to do this, though. Needs more thought.

Reply:
gcc 10 isn't out yet, could ask Wakely very nicely to stick the variable template and the base class into its own header (right now they're both in `<ranges>`: https://github.com/gcc-mirror/gcc/blob/master/libstdc++-v3/include/std/ranges#L66-L69).

`<version>` exists now so could check the feature-test macro from there.

Reply:
Couldn't we just forward declare `std::ranges::enable_view` and `std::ranges::view_base` and do the manual overload that I did in my issue? So ranges keeps it's base classes, but enables them as views?

I think you also forward declared std::span in ranges without actually pulling it in.

Reply:
They can, but on MSVC, you need [`/Zc:externConstexpr`](https://docs.microsoft.com/en-us/cpp/build/reference/zc-externconstexpr?view=vs-2019).

```c++
template<class>
extern const bool enable_view;
```
 
Later...

```c++
template <class T>
inline constexpr bool enable_view = ...;
```

Reply:
... however, it isn't enough to merely forward-declare these variable templates as @BRevzin is doing here. It will lead to hard failures if the definition is not available at the point it is ODR used, as shown here: https://godbolt.org/z/Ewuq4k.

Reply:
That's weird, because all the tests compile for me even though only one of them includes `<ranges>`...

Reply:
You can't forward-declare variable templates; these are definitions.

Reply:
> That's weird, because all the tests compile for me even though only one of them includes <ranges>...

I'm going to guess that these are getting pulled in from _somewhere_ besides `<ranges>`, but that it's not something we can or should be relying on.

Reply:
@ericniebler @CaseyCarter Ok tracked that down. `<algorithm>` includes `<bits/ranges_algo.h>` includes `<bits/ranges_algobase.h>` includes `<ranges>`. So that's probably how it's getting in (although really, every unit test in range-v3 includes `<algorithm>`?)

I guess this direction is a non-starter then.